### PR TITLE
Fix pytorch eval for fp8

### PR DESCRIPTION
**Pull request by j316chuck**
# What does this PR do?
Turn off te fp8 autocast for eval. This logic works around the incompatability where composer eval batches can be of any arbitrary shape whereas TE requires that eval batches are divisible by 16.

# What issue(s) does this change relate to?
https://databricks.atlassian.net/browse/GRT-3023

# Unit Tests
- H100: `WORLD_SIZE=1 EXTRA_ARGS='-k test_amp_fp8_eval_casts_to_bf16' make test-dist-gpu` ✅ 
- A100: `WORLD_SIZE=1 EXTRA_ARGS='-k test_amp_fp8_eval_casts_to_bf16' make test-dist-gpu` ✔️ (skipped) 

# Regression Tests
- Before: `fp8-llama3-8b-metamath-4ep-3AOZfj` 🔴 (fails on model eval)
- After: `fp8-llama3-8b-metamath-4ep-tIKVb1` ✅ (works on model eval)

# Along for the ride
- Fix spurious [test failure](https://github.com/mosaicml/composer/actions/runs/9885666578/job/27304504456?pr=3454#step:6:1276) in HF tokenizer by adding retry logic to test fixture

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -3479,13 +3485,20 @@ def _eval_loop(
                                         )[0]
 
                             self.engine.run_event(Event.EVAL_BEFORE_FORWARD)
+                            if self.state.precision is Precision.AMP_FP8:
+                                import transformer_engine.pytorch as te
+                                log.info("Disabling AMP FP8 for eval loop")
+                                fp8_ctx = te.fp8_autocast(enabled=False)
+                            else: 
+                                fp8_ctx = contextlib.nullcontext()
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:3493]: maybe make helper fn? there are other parts in codebase (eg DPO callback) that require a pass which would want this context

File: composer/core/precision.py [hunk 2]
```diff
@@ -36,6 +36,25 @@ class Precision(StringEnum):
     AMP_FP8 = 'amp_fp8'
 
 
+def get_fp8_precision_context(
```
Comment (mvpatel2000) [inline composer/core/precision.py:39]: can we absorb into existing precision context fn?

File: composer/trainer/trainer.py [hunk 3]
```diff
@@ -3470,13 +3477,19 @@ def _eval_loop(
...
                                 self.state.deepspeed_enabled,
                             ):
-                                self.state.outputs = self._original_model.eval_forward(self.state.batch)
+                                # Disabling FP8 in eval mode and defaulting to BF16.
+                                # This is because FP8 in TE requires all eval data sizes to be divisible by 16 which does not hold for all evaluation datasets.
+                                # See https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/fp8_primer.html for more info.
+                                with get_fp8_precision_context(
+                                    fp8_autocast_enabled=False,
+                                    precision_config=None,
+                                ):
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:3491]: does this run in fp32 or b16?

File: composer/core/precision.py (hunk 2 above)
Comment (j316chuck) [reply composer/core/precision.py]: 👍

File: composer/trainer/trainer.py (hunk 3 above)
Comment (j316chuck) [reply composer/trainer/trainer.py]: bf16. see run here: `fp8-llama3-8b-metamath-4ep-Ok7FKv`

**Review by mvpatel2000 — APPROVED** [review]
LGTM! Ty :). Can you please ensure unit test runs on an interactive instance?

**Pull request merged**